### PR TITLE
Added explicit decencies for surefile-junit4 in order to fix testing in offline mode. 

### LINKED ISSUE #3819: dependency:go-offline fails to pull in required surefire testing dependency.
Thanks for stopping by to let us know something could be better!

Please include as much information as possible:

#### Environment details

- OS: Linux
- Java version: java8
- google-cloud-java version(s):

#### Steps to reproduce
0. Start with a completely clean environment. 
1. Run a command that builds the project without running tests: mvn install -DskipTests
2. Pull down the dependencies:  dependency:go-offline
3. Trying running the tests off line: mvn --fail-at-end -o verify

#### Stacktrace
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project google-cloud-core: Unable to generate classpath: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException: Missing:
[ERROR] ----------
[ERROR] 1) org.apache.maven.surefire:surefire-junit4:jar:2.19.1
[ERROR]
[ERROR] Try downloading the file manually from the project website.
[ERROR]
[ERROR] Then, install it using the command:
[ERROR] mvn install:install-file -DgroupId=org.apache.maven.surefire -DartifactId=surefire-junit4 -Dversion=2.19.1 -Dpackaging=jar -Dfile=/path/to/file
[ERROR]
[ERROR] Alternatively, if you host your own repository you can deploy the file there:
[ERROR] mvn deploy:deploy-file -DgroupId=org.apache.maven.surefire -DartifactId=surefire-junit4 -Dversion=2.19.1 -Dpackaging=jar -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR]
[ERROR] Path to dependency:
[ERROR] 1) dummy:dummy:jar:1.0
[ERROR] 2) org.apache.maven.surefire:surefire-junit4:jar:2.19.1
[ERROR]
[ERROR] ----------
[ERROR] 1 required artifact is missing.
[ERROR]
[ERROR] for artifact:
[ERROR] dummy:dummy:jar:1.0
[ERROR]
[ERROR] from the specified remote repositories:
[ERROR] central (http://repo.maven.apache.org/maven2, releases=true, snapshots=false)
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

Thanks!

### PR DESCRIPTION
**Pull request by silverdev**
Fixes #3819 (it's a good idea to open an issue first for context and/or discussion)

### CODE REVIEW

File: google-cloud-clients/google-cloud-spanner/pom.xml [hunk 1]
```diff
@@ -191,5 +198,10 @@
       <artifactId>json</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.maven.surefire</groupId>
+      <artifactId>surefire-junit4</artifactId>
+      <version>2.12.4</version>
```
[the remark this answers is not in the document]
Comment (silverdev) [reply google-cloud-clients/google-cloud-spanner/pom.xml]: I don't understand what you are saying. For example in the file linked bellow the maven-failsafe-plugin is at 2.19.1. This means that the surefire-junit4 needs to be of the same version. Are you saying that I can just leave of the <version>2.19.1</version> tag on surefire-junit4 or that I can use some type of annotation to pull it in from failsafe?

https://github.com/googleapis/google-cloud-java/blob/6ca9219c127c65ec6755b29988fb9452645ff529/google-cloud-clients/google-cloud-bigtable/pom.xml#L122

[the remark this answers is not in the document]
Comment (chingor13) [reply google-cloud-clients/google-cloud-spanner/pom.xml]: Maven project files can inherit from their parents. All of the google-cloud-x under google-cloud-clients share a parent named `google-cloud-clients`. If you move these changes to the parent pom, then all the children will inherit the changes - so we don't need to change them in X places.

Additionally, you can separate the configuration for dependencies into a dependencyManagement section. 

So in the `google-cloud-clients/pom.xml` you could have something like:
```xml
<project>
...snip
  <dependencies>
    <dependency>
      <groupId>org.apache.maven.surefire</groupId>
      <artifactId>surefire-junit4</artifactId>
      <scope>test</scope>
    <dependency>
  </dependencies>
  <dependencyManagement>
    <dependencies>
      ...snip
      <dependency>
        <groupId>org.apache.maven.surefire</groupId>
        <artifactId>surefire-junit4</artifactId>
        <version>2.12.4</version>
      <dependency>
      ...snip
```

[the remark this answers is not in the document]
Comment (silverdev) [reply google-cloud-clients/google-cloud-spanner/pom.xml]: So I tried defining it the google-cloud-clients/pom.xml. However failsafe is defined library-by-library and has at lest three different versions. I don't know the reasons for this or its implications but if found it easier to stick with what the version the test owners chose. I would recommend a clean up but that is outside of the scope of this safe PR. I'm telling maven to pull in the dependency it was pulling in implicitly at test time. Trying to make all twelve use the same version might break things.

**Comment by chingor13** [comment]
All the versions of surefire should be the same within google-cloud-java. The version that other projects is irrelevant as it is a **_test_** dependency - only used for building and running the tests this package.

**Comment by codahale** [comment]
This change added `org.apache.maven.surefire:surefire-junit4:2.19.1` as a compile dependency to most of the GCP SDKs, which makes it impossible to depend on those SDKs and run tests in Maven.

**Comment by silverdev** [comment]
Hello Coda,
I don't know why you mean by you can't run tests in Maven. I had no trouble
using maven to build and run these libraries in a number of ways and it
passes our tests (run in Maven). The reason I added
the org.apache.maven.surefire:surefire-junit4 dependency is to fix Maven's
ability to run the tests in offline mode. I'm no expert with Maven so if
know a better way to add said dependencies it would works for me. I just
need to make sure that it is still downloaded by "mvn
dependency:go-offline" so it can be run with "mvn -o verify". As far as I
can tell surefire-junit4 if requeres to run the tests on the library but
not pulled in by "mvn dependency:go-offline" unless it's listed as a
dependency. Is there a different place test dependencies are normally
included?

Sam Laane

**Comment by silverdev** [comment]
I see the problem and and more then happy to help you get a fix it. I have
a commit on my fork that moves surefire to be in the plugin section. Does
it work for your use case? If so I'll work the the maintainers to get it
merged to head.

https://github.com/silverdev/google-cloud-java/commit/35feb8fb115f41d6ad55298182b5cc4093db984d

**Comment by codahale** [comment]
That should work, yes.

**Pull request merged**